src/librustc_infer/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -2,229 +2,300 @@
+                                    self_ty.span.shrink_to_hi(),
+                                    "this `impl` introduces an implicit `'static` requirement, \
+                                     consider changing it",
+                                    " + '_".to_string(),
+                                    Applicability::MaybeIncorrect,
+                                );
+                                suggested = true;
                             }
-                            TyKind::TraitObject(_, lt) => match lt.name {
-                                LifetimeName::ImplicitObjectLifetimeDefault => {
-                                    err.span_suggestion_verbose(
-                                        fn_return.span.shrink_to_hi(),
-                                        &format!(
-                                            "{declare} trait object {captures}, {explicit}",
-                                            declare = declare,
-                                            captures = captures,
-                                            explicit = explicit,
-                                        ),
-                                        plus_lt.clone(),
-                                        Applicability::MaybeIncorrect,
-                                    );
-                                }
-                                name if name.ident().to_string() != lifetime_name => {
-                                    // With this check we avoid suggesting redundant bounds. This
-                                    // would happen if there are nested impl/dyn traits and only
-                                    // one of them has the bound we'd suggest already there, like
-                                    // in `impl Foo<X = dyn Bar> + '_`.
-                                    err.span_suggestion_verbose(
-                                        lt.span,
-                                        &format!("{} trait object's {}", consider, explicit_static),
-                                        lifetime_name.clone(),
-                                        Applicability::MaybeIncorrect,
-                                    );
-                                    err.span_suggestion_verbose(
-                                        param_info.param_ty_span,
-                                        add_static_bound,
-                                        param_info.param_ty.to_string(),
-                                        Applicability::MaybeIncorrect,
-                                    );
-                                }
-                                _ => {}
-                            },
-                            _ => {}
                         }
                     }
-                    err.emit();
-                    return Some(ErrorReported);
                 }
             }
         }
-        None
+        suggested
+    }
```

Comment:
@nikomatsakis could you take a look at this and let me know if there's a better way of doing this? Conceptually, what I want is to _only_ get the `impl` blocks that contributed to the currently seen `'static` obligation, regardless if it is from `impl Foo for dyn Bar {}` or `impl dyn Bar {}` or `impl Foo for Box<dyn Bar> {}`. Instead, we look at the `impl`s and filter until we have `impl Foo for dyn Bar {}`. I ain't happy about this approach, but it seems like a reasonable stopgap.